src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,696 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::Result;
+use std::{str, f64};
+use std::cmp::{Ord, Ordering, PartialEq, PartialOrd};
+use std::collections::BTreeMap;
+use std::result::Result as SResult;
+use serde_json::{self, Value};
+use std::io::{Read, Write};
+use util::codec::number::{NumberDecoder, NumberEncoder};
+use byteorder::{ReadBytesExt, WriteBytesExt};
+use serde::ser::{Serialize, Serializer, SerializeTuple, SerializeMap};
+
+const TYPE_CODE_OBJECT: u8 = 0x01;
+const TYPE_CODE_ARRAY: u8 = 0x03;
+const TYPE_CODE_LITERAL: u8 = 0x04;
+const TYPE_CODE_I64: u8 = 0x09;
+const TYPE_CODE_DOUBLE: u8 = 0x0b;
+const TYPE_CODE_STRING: u8 = 0x0c;
+
+#[allow(dead_code)]
+const PRECEDENCE_BLOB: i32 = -1;
+#[allow(dead_code)]
+const PRECEDENCE_BIT: i32 = -2;
+#[allow(dead_code)]
+const PRECEDENCE_OPAQUE: i32 = -3;
+#[allow(dead_code)]
+const PRECEDENCE_DATETIME: i32 = -4;
+#[allow(dead_code)]
+const PRECEDENCE_TIME: i32 = -5;
+#[allow(dead_code)]
+const PRECEDENCE_DATE: i32 = -6;
+const PRECEDENCE_BOOLEAN: i32 = -7;
+const PRECEDENCE_ARRAY: i32 = -8;
+const PRECEDENCE_OBJECT: i32 = -9;
+const PRECEDENCE_STRING: i32 = -10;
+const PRECEDENCE_NUMBER: i32 = -11;
+const PRECEDENCE_NULL: i32 = -12;
+
+const JSON_LITERAL_NIL: u8 = 0x00;
+const JSON_LITERAL_TRUE: u8 = 0x01;
+const JSON_LITERAL_FALSE: u8 = 0x02;
+
+const LENGTH_TYPE: usize = 1;
+const LENGTH_LITERAL: usize = 1;
+const LENGTH_U16: usize = 2;
+const LENGTH_U32: usize = 4;
+const LENGTH_I64: usize = 8;
+const LENGTH_KEY_ENTRY: usize = LENGTH_U32 + LENGTH_U16;
+const LENGTH_VALUE_ENTRY: usize = LENGTH_TYPE + LENGTH_U32;
+
+const ERR_CONVERT_FAILED: &str = "Can not covert from ";
+
+// Json implement type json used in tikv, it specifies the following
+// implementations:
+// 1. Serialize `json` values into binary representation, and reading values
+//  back from the binary representation.
+// 2. Serialize `json` values into readable string representation, and reading
+// values back from string representation.
+// The binary Json format from MySQL 5.7 is in the following link:
+// https://github.com/mysql/mysql-server/blob/5.7/sql/json_binary.h#L52
+// The only difference is that we use large `object` or large `array` for
+// the small corresponding ones. That means in our implementation there
+// is no difference between small `object` and big `object`, so does `array`.
+#[derive(Clone, Debug)]
+pub enum Json {
+    Object(BTreeMap<String, Json>),
+    Array(Vec<Json>),
+    Literal(u8),
```

Comment:
I guess it would be more clear if we keep consistent with `tidb`(https://github.com/pingcap/tidb/pull/3248)? 
And  could we discuss this issue in the (https://github.com/pingcap/tikv/pull/1874)? @andelf